src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -0,0 +1,382 @@
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.memtable.TrieMemtable;
+import org.apache.cassandra.db.tries.InMemoryTrie;
+import org.apache.cassandra.db.tries.Trie;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.analyzer.AbstractAnalyzer;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.PrimaryKeys;
+import org.apache.cassandra.index.sai.utils.RangeIterator;
+import org.apache.cassandra.index.sai.utils.TypeUtil;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.cassandra.utils.bytecomparable.ByteSourceInverse;
+
+/**
+ * This is an in-memory index using the {@link InMemoryTrie} to store a {@link ByteComparable}
+ * representation of the indexed values. Data is stored on-heap or off-heap and follows the
+ * settings of the {@link TrieMemtable} to determine where.
+ *
+ *
+ */
+public class TrieMemoryIndex
+{
+    private static final Logger logger = LoggerFactory.getLogger(TrieMemoryIndex.class);
+    private static final int MINIMUM_QUEUE_SIZE = 128;
+    private static final int MAX_RECURSIVE_KEY_LENGTH = 128;
+
+    private final IndexContext indexContext;
+    private final InMemoryTrie<PrimaryKeys> data;
+    private final PrimaryKeysReducer primaryKeysReducer;
+    private final AbstractAnalyzer.AnalyzerFactory analyzerFactory;
+    private final AbstractType<?> validator;
+    private final boolean isLiteral;
+    private final Object writeLock = new Object();
+
+    private ByteBuffer minTerm;
+    private ByteBuffer maxTerm;
+
+    private static final FastThreadLocal<Integer> lastQueueSize = new FastThreadLocal<Integer>()
+    {
+        protected Integer initialValue()
+        {
+            return MINIMUM_QUEUE_SIZE;
+        }
+    };
+
+
+    public TrieMemoryIndex(IndexContext indexContext)
+    {
+        this.indexContext = indexContext;
+        this.data = new InMemoryTrie<>(TrieMemtable.BUFFER_TYPE);
+        this.primaryKeysReducer = new PrimaryKeysReducer();
+        // The use of the analyzer is within a synchronized block so can be considered thread-safe
+        this.analyzerFactory = indexContext.getIndexAnalyzerFactory();
+        this.validator = indexContext.getValidator();
+        this.isLiteral = TypeUtil.isLiteral(validator);
+    }
+
+    /**
+     * Adds a index value to the in-memory index
+     *
+     * @param key partition key for the indexed value
+     * @param clustering clustering for the indexed value
+     * @param value indexed value
+     * @return amount of heap allocated by the new value
+     */
+    public long add(DecoratedKey key, Clustering<?> clustering, ByteBuffer value)
+    {
+        synchronized (writeLock)
```

Comment:
nit: Any reason not to just make `add()` synchronized and do away w/ this explicit locking object? (We're just doing this to serialize writes to `InMemoryTrie`, given we aren't doing TPC/thread-per-memtable shard?)

Reply:
We have done this in a follow on ticket that does sharding of the index to memtable shard and has this change in it. I'll do this here for now since you have raised it.